include/hipSYCL/sycl/libkernel/builtins.hpp

```diff
@@ -507,7 +511,10 @@ using ulonglong16 = vec<unsigned long long, 16>;
   typename detail::builtin_type_traits<T>::alternative_data_type<int64_t> name(\
       T a) noexcept {                                                          \
     if constexpr (std::is_arithmetic_v<T>) {                                   \
-      return impl_name(detail::data_element(a, 0));                            \
+      return static_cast<                                                      \
+          typename detail::builtin_type_traits<T>                              \
+                         ::alternative_data_type<int64_t>>                     \
+                             (impl_name(detail::data_element(a, 0)));          \
```

Comment:
Are the changes in this file related?

Reply:
Yes, and I only half understand why. Without the explicit casts, the compiler complains when `T` is a `sycl::vec`: in this case, it does not know how to convert from the return type of `impl_name` (for example `double`) to `T` (for example `sycl::vec<double, 2>`). The `vec(const Args &...args)` constructor is now not applicable anymore because one `double` is not enough for a `sycl::vec` with two elements and the `vec(const T &value)` constructor is not applicable because it is marked explicit. 

The funny thing is that this happens in the `if constexpr` branch that will never be executed for `T == sycl::vec` because `sycl::vec` is not an arithmetic type. However, the C++ standard apparently requires both branches in an `if constexpr` to be compilable if the `if constexpr` is not inside a template (not sure what the reason for this is). 

And by explicitly casting here, the `explicit` constructor of the `vec` class is now a possible option and the compiler is happy (even though it will never call this constructor).